File: sprint8/By lessons/lesson5code2/lesson5code2.kt
fun prefix_function(s: String): IntArray {
    // Функция возвращает массив длины |s|
    val π = IntArray(s.length) { 0 }
    for (i in 1 until s.length) {
        var k = π[i - 1]
        while (k > 0 && s[k] != s[i]) {
            k = π[k - 1]
        }
        if (s[k] == s[i]) {
            k++
        }
        π[i] = k
    }
    return π
}

File: sprint8/By lessons/lesson5code2/lesson5code2.cpp
vector<int> prefix_function(string s) {
    int N = s.length();
    vector<int> π(N, 0);
    for (int i = 1; i < N; i++) {
        int k = π[i - 1];
        while (k > 0 && s[k] != s[i]) {
            k = π[k - 1];
        }
        if (s[k] == s[i]) {
            k += 1;
        }
        π[i] = k;
    }
    return π;
}
